iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,451 @@
+            let tree_len = tree.len();
+
+            // check if the leaf node of the tree is full
+            if tree[0].len() == degree {
+                // if so, iterate through nodes
+                for i in 0..tree_len {
+                    // if we encounter any nodes that are not full, break
+                    if tree[i].len() < degree {
+                        break;
+                    }
+
+                    // in this case we have a full set of links & we are
+                    // at the top of the tree. Time to make a new layer.
+                    if i == tree_len - 1 {
+                        tree.push_back(vec![]);
+                    }
+
+                    // create node, keeping the cid
+                    let links = std::mem::replace(&mut tree[i], Vec::new());
+                    let (cid, bytes, len) = TreeNode::Stem(links).encode()?;
+                    yield (cid, bytes);
+
+                    // add cid to parent node
+                    tree[i+1].push((cid, len));
+                }
+                // at this point the tree will be able to recieve new links
+                // without "overflowing", aka the leaf node and stem nodes
+                // have fewer than `degree` number of links
+            }
+
+            // now that we know the tree is in a "healthy" state to
+            // recieve more links, add the link to the tree
+            let (cid, bytes, len) = TreeNode::Leaf(chunk).encode()?;
+            tree[0].push((cid, len));
+            yield (cid, bytes);
+            // at this point, the leaf node may have `degree` number of
+            // links, but no other stem node will
+        }
+
+        // our stream had 1 chunk that we have already yielded
+        if tree.len() == 1 && tree[0].len() == 1 {
+            return
+        }
+
+        // clean up, aka yield the rest of the stem nodes
+        // since all the stem nodes are able to recieve links
+        // we don't have to worry about "overflow"
+        while let Some(links) = tree.pop_front() {
+            let (cid, bytes, len) = TreeNode::Stem(links).encode()?;
+            yield (cid, bytes);
+
+            if let Some(front) = tree.front_mut() {
+                front.push((cid, len));
+            } else {
+                // final root, nothing to do
+            }
+        }
+    }
+}
+
+fn create_unixfs_node_from_links(links: Vec<(Cid, u64)>, is_root: bool) -> Result<UnixfsNode> {
+    let links = links
+        .into_iter()
+        .map(|(cid, len)| dag_pb::PbLink {
+            hash: Some(cid.to_bytes()),
+            name: None,
+            tsize: Some(len as u64),
+        })
+        .collect();
+
+    // PBNode.Data
+    let inner = unixfs_pb::Data {
+        r#type: DataType::File as i32,
+        ..Default::default()
+    };
+
+    // create PBNode
+    let outer = encode_unixfs_pb(&inner, links)?;
+
+    if is_root {
+        return Ok(UnixfsNode::Directory(Node { inner, outer }));
```

Comment:
If this is a directory, the links should have names, but I don't think that is the right type for the root here, unless I am missing sth

Reply:
Oof, this is dead code that should have been removed. Thank you for catching.